### ISSUE TITLE
Trino's lambda notation results in incorrect reference tracking

### ISSUE THREAD

**Issue by namelessjon** [sqlfluff/sqlfluff#6141]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Trino introduced the [TRANSFORM](https://trino.io/docs/current/functions/array.html#transform) SQL function, which accepts a lambda function as a parameter. The lambda notation includes an arrow sign (->) but the results of this are not interpreted correctly, leading to "Reference 'x.example' refers to table/view not found in the FROM clause or found in ancestor statement."

### Expected Behaviour

SQLfluff should parse this sql correctly with no errors related to table/view not found:
```sql
SELECT
    a_column,
    TRANSFORM(array_col, x -> x.example) AS array_col_example
FROM example_table
```

### Observed Behaviour

```
L:   2 | P:   5 | RF03 | Unqualified reference 'a_column' found in single table
                       | select. [references.consistent]
L:   3 | P:  15 | RF03 | Unqualified reference 'array_col' found in single table
                       | select. [references.consistent]
L:   3 | P:  26 | RF03 | Unqualified reference 'x' found in single table select.
                       | [references.consistent]
L:   3 | P:  31 | RF01 | Reference 'x.example' refers to table/view not found in
                       | the FROM clause or found in ancestor statement.
                       | [references.from]
L:   3 | P:  31 | RF03 | Qualified reference 'x.example' found in single table
                       | select which is inconsistent with previous references.
                       | [references.consistent]
```

### How to reproduce

use the example sql, and `sqlfluff lint <file>.sql`

### Dialect

trino

### Version

* sqlfluff 3.1.1
* python 3.12.3, 3.11.7

### Configuration

vanilla

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by namelessjon** [comment]
I think to resolve this, the trino dialect needs to parse lambda expressions in a similar manner to #6136, i.e. as a lambda expression, not as a `column_reference`, `binary operator ->` as https://github.com/sqlfluff/sqlfluff/blob/main/test/fixtures/dialects/trino/regexp_replace_with_lambda.yml#L26-L31 suggests it is.

**Comment by namelessjon** [comment]
No, some playing with the grammar suggests that isn't enough, there needs to something which links the `x` on the LHS of the `->` with the `x` on the RHS.

**Comment by namelessjon** [comment]
While I'm still willing to submit a PR, I think I'm going to need help on where to start :)
